File: frontend/spec-story.js
function escapeHtml(text) {
  const div = document.createElement("div");
  div.textContent = text ?? "";
  return div.innerHTML;
}

function tsFromPath() {
  const parts = window.location.pathname.split("/").filter(Boolean);
  const idx = parts.indexOf("specs");
  if (idx >= 0 && parts[idx + 1]) return decodeURIComponent(parts[idx + 1]);
  return null;
}

function isEmbedMode() {
  return new URLSearchParams(window.location.search).get("embed") === "1";
}

function isDebugMode() {
  return new URLSearchParams(window.location.search).get("debug") === "1";
}

function wrapTable(html) {
  return `<div class="table-scroll">${html}</div>`;
}

function badge(type, label) {
  return `<span class="badge ${type}">${escapeHtml(label)}</span>`;
}

function renderEntryPoints(entries) {
  if (!entries?.length) return "<p>No entry points defined.</p>";
  const rows = entries
    .map((e) => {
      const typeBadge =
        e.type === "sbi" ? badge("sbi", "SBI") : e.type === "procedure" ? badge("proc", "Procedure") : badge("neighbor", e.type);
      const detail =
        e.type === "sbi"
          ? `<code>${escapeHtml(e.method || "")} ${escapeHtml(e.path || "")}</code><br>${escapeHtml(e.operationId || e.name || "")}`
          : `${escapeHtml(e.name || "")}${e.clause ? ` <em>(§${escapeHtml(e.clause)})</em>` : ""}`;
      return `<tr>
        <td>${typeBadge}</td>
        <td>${detail}</td>
        <td>${escapeHtml(e.consumer || e.direction || "")}</td>
        <td>${escapeHtml(e.description || e.trigger || "")}${e.refSpec ? ` <em>(TS ${escapeHtml(e.refSpec)})</em>` : ""}</td>
      </tr>`;
    })
    .join("");
  return wrapTable(
    `<table class="story-table"><thead><tr><th>Type</th><th>Endpoint / name</th><th>Actor</th><th>Description</th></tr></thead><tbody>${rows}</tbody></table>`
  );
}

function renderRelatedStory(related) {
  if (!related?.length) return "";
  const links = related
    .map(
      (r) =>
        `<li><a href="/specs/${encodeURIComponent(r.ts)}">TS ${escapeHtml(r.ts)}</a> — ${escapeHtml(r.role || "")}</li>`
    )
    .join("");
  return `<div class="related-story"><h3 class="related-story-title">Related in the 512 story</h3><ul class="related-story-list">${links}</ul></div>`;
}

function readinessChip(label, on) {
  return `<span class="ready-chip ${on ? "on" : "off"}">${escapeHtml(label)}</span>`;
}

function renderElementReadinessItem(item) {
  if (!item) return "";
  const chips = [
    readinessChip("spec", item.hasSpec),
    readinessChip("yaml", item.hasYaml),
    readinessChip("story", item.hasStory),
  ].join("");
  const badgeLabel = item.ready ? "ready" : item.state === "partial" ? "partial" : "gap";
  const badgeType = item.ready ? "in-spec" : item.state === "partial" ? "sbi" : "neighbor";
  const specLink = item.primarySpec
    ? ` <a class="element-spec-link" href="${escapeHtml(item.platformUrl || `/specs?ts=${item.primarySpec}`)}" target="_blank" rel="noopener">TS ${escapeHtml(item.primarySpec)}</a>`
    : "";
  return `${badge(badgeType, badgeLabel)}${specLink}<div class="ready-chips">${chips}</div>`;
}

function renderElements(elements, readiness, nfGaps) {
  if (!elements?.length) return "<p>No elements defined.</p>";
  const debug = isDebugMode();
  const legend = window.FlowRenderer?.renderJourneyLegend
    ? window.FlowRenderer.renderJourneyLegend()
    : "";
  const gapsTable =
    debug && nfGaps?.items?.length && window.FlowRenderer?.renderNfGapsTable
      ? window.FlowRenderer.renderNfGapsTable(nfGaps, { compact: true })
      : "";

  return `${legend}${gapsTable}<div class="element-grid">${elements
    .map((el) => {
      const journeyClass = el.inSpec ? "journey-focus" : "journey-related";
      const primary = el.primarySpec || "";
      const specLink = primary
        ? `<a class="element-spec-link" href="/specs/${encodeURIComponent(primary)}" target="_blank" rel="noopener">TS ${escapeHtml(primary)}</a>`
        : "";
      const readinessItem = readiness?.items?.find((r) => r.nf === el.nf);
      const readinessHtml = debug && readinessItem ? renderElementReadinessItem(readinessItem) : "";
      return `<div class="element-card ${journeyClass}">
      <h3>${escapeHtml(el.nf)}</h3>
      ${specLink}
      ${readinessHtml}
      <p>${escapeHtml(el.role)}</p>
    </div>`;
    })
    .join("")}</div>`;
}

function renderInteractions(interactions, operations) {
  if (!interactions?.length && !operations?.length) return "<p>No interactions defined.</p>";
  const rows = (interactions || [])
    .map(
      (ix) => `<tr>
      <td>${escapeHtml(ix.from)} → ${escapeHtml(ix.to)}</td>
      <td><code>${escapeHtml(ix.via || "")}</code></td>
      <td>${escapeHtml(ix.payload || ix.response || ix.description || "")}${ix.refSpec ? ` <em>(TS ${escapeHtml(ix.refSpec)})</em>` : ""}</td>
    </tr>`
    )
    .join("");
  const opRows = (operations || [])
    .map(
      (op) => `<tr>
      <td>${escapeHtml(op.nf)} → PCF</td>
      <td><code>${escapeHtml(op.method)} ${escapeHtml(op.path)}</code></td>
      <td><a href="/?yaml=${encodeURIComponent(op.sourceFile)}">${escapeHtml(op.operationId)}</a> — ${escapeHtml(op.summary)}</td>
    </tr>`
    )
    .join("");
  return wrapTable(
    `<table class="story-table"><thead><tr><th>Link</th><th>Via</th><th>Detail</th></tr></thead><tbody>${rows}${opRows}</tbody></table>`
  );
}

function nfMapFromReadiness(readiness) {
  const map = {};
  for (const item of readiness?.items || []) {
    map[item.nf] = { storyUrl: item.storyUrl };
  }
  return map;
}

function renderStoryGaps(gaps) {
  if (!gaps?.items?.length) return "";
  const meta = gaps.meta || {};
  const rows = gaps.items
    .map((item) => {
      const pdf = item.hasPdf ? badge("in-spec", item.pdfVersion || "PDF") : badge("neighbor", "no PDF");
      const yaml = item.hasYaml ? badge("sbi", "YAML") : badge("neighbor", "ref only");
      const clauses = item.hasClauses ? `${item.clauseCount} clauses` : "—";
      return `<tr>
        <td><span class="gap-priority">${escapeHtml(item.priority || "")}</span> <a href="${escapeHtml(item.storyUrl)}"><strong>TS ${escapeHtml(item.tsNumber)}</strong></a></td>
        <td>${escapeHtml(item.role || "")}</td>
        <td>${pdf} ${yaml}</td>
        <td>${escapeHtml(clauses)}</td>
        <td>${item.ready ? badge("in-spec", "stitched") : badge("neighbor", "gap")}</td>
      </tr>`;
    })
    .join("");
  const headline = meta.complete
    ? "All supporting specs are available — the 512 story is fully stitched."
    : `${meta.ready} of ${meta.total} supporting specs ready — import remaining PDFs to close knowledge gaps.`;
  return `<p class="gaps-headline">${escapeHtml(headline)}</p>${wrapTable(
    `<table class="story-table gaps-table"><thead><tr><th>Spec</th><th>Role in 512 story</th><th>Sources</th><th>PDF index</th><th>Status</th></tr></thead><tbody>${rows}</tbody></table>`
  )}`;
}

function renderSpecReadiness(r) {
  if (!r) return "";
  const chips = [
    readinessChip("spec", r.hasSpec),
    readinessChip("yaml", r.hasYaml),
    readinessChip("story", r.hasStory),
  ].join("");
  const label = r.ready ? "ready" : r.state === "partial" ? "partial" : "gap";
  const badgeType = r.ready ? "in-spec" : r.state === "partial" ? "sbi" : "neighbor";
  return `<div class="spec-readiness spec-readiness-${r.state}">${badge(badgeType, label)} <span class="spec-readiness-label">TS ${escapeHtml(r.tsNumber)} readiness</span><div class="ready-chips">${chips}</div></div>`;
}

function renderOperations(operations) {
  if (!operations?.length) return "<p>No operations in corpus.</p>";
  const rows = operations
    .map(
      (op) => `<tr>
      <td><code>${escapeHtml(op.method)}</code></td>
      <td><code>${escapeHtml(op.path)}</code></td>
      <td><a href="/?yaml=${encodeURIComponent(op.sourceFile)}">${escapeHtml(op.operationId)}</a></td>
      <td>${escapeHtml(op.summary)}</td>
    </tr>`
    )
    .join("");
  return wrapTable(
    `<table class="story-table"><thead><tr><th>Method</th><th>Path</th><th>Operation</th><th>Summary</th></tr></thead><tbody>${rows}</tbody></table>`
  );
}

function renderReferencedBy(files) {
  if (!files?.length) return "<p>No cross-references in corpus.</p>";
  return `<ul class="ref-list">${files
    .map((f) => `<li><a href="/?yaml=${encodeURIComponent(f)}">${escapeHtml(f)}</a></li>`)
    .join("")}</ul>`;
}

function renderClauses(clauses, highlightNums, ts) {
  if (!clauses?.clauses?.length) {
    const cmd = ts ? `python scripts/extract_spec_pdf.py ${ts}` : "python scripts/extract_spec_pdf.py <ts>";
    return `<p>PDF clauses not extracted yet. Run <code>${escapeHtml(cmd)}</code>.</p>`;
  }
  const all = clauses.clauses;
  const highlights =
    highlightNums?.length > 0
      ? all.filter((c) => highlightNums.some((h) => c.number === h || c.number.startsWith(`${h}.`)))
      : [];
  const highlightBlock = highlights.length
    ? `<div class="clause-highlights"><h3 class="clause-highlights-title">Key clauses for this story</h3><ul class="clause-list">${highlights
        .map((c) => `<li><strong>${escapeHtml(c.number)}</strong> ${escapeHtml(c.title)}</li>`)
        .join("")}</ul></div>`
    : "";
  const items = all.map((c) => `<li><strong>${escapeHtml(c.number)}</strong> ${escapeHtml(c.title)}</li>`).join("");
  return `${highlightBlock}<p class="meta-chip">${escapeHtml(clauses.clauseCount || all.length)} clauses from ${escapeHtml(clauses.sourceFile || "PDF")} ${escapeHtml(clauses.version || "")}</p><div class="clause-scroll"><ul class="clause-list">${items}</ul></div>`;
}

function updatePlatformLinks(ts) {
  const platform = document.getElementById("bc-platform");
  if (platform) platform.href = `/specs?ts=${encodeURIComponent(ts)}&view=story`;
}

function initTocSpy() {
  const tocLinks = document.querySelectorAll(".toc-link");
  const scrollRoot = document.querySelector(".story-scroll");
  if (!tocLinks.length || !scrollRoot) return;

  const sectionTop = (el) => {
    const rootRect = scrollRoot.getBoundingClientRect();
    const elRect = el.getBoundingClientRect();
    return elRect.top - rootRect.top + scrollRoot.scrollTop;
  };

  tocLinks.forEach((link) => {
    link.addEventListener("click", (e) => {
      e.preventDefault();
      const id = link.getAttribute("href")?.slice(1);
      const target = id ? document.getElementById(id) : null;
      if (target) {
        scrollRoot.scrollTo({ top: sectionTop(target) - 8, behavior: "smooth" });
      }
    });
  });

  const sections = [...tocLinks]
    .map((link) => {
      const id = link.getAttribute("href")?.slice(1);
      const el = id ? document.getElementById(id) : null;
      return el ? { link, el } : null;
    })
    .filter(Boolean);

  const observer = new IntersectionObserver(
    (entries) => {
      const visible = entries.filter((e) => e.isIntersecting).sort((a, b) => b.intersectionRatio - a.intersectionRatio);
      if (!visible.length) return;
      const id = visible[0].target.id;
      tocLinks.forEach((l) => l.classList.toggle("active", l.getAttribute("href") === `#${id}`));
    },
    { root: scrollRoot, rootMargin: "-10% 0px -55% 0px", threshold: [0, 0.25, 0.5] }
  );

  sections.forEach((s) => observer.observe(s.el));
}

function initTopologyGraph(topology) {
  const el = document.getElementById("interaction-graph");
  if (!el || typeof vis === "undefined" || !topology?.nodes?.length) return;

  const focusNf = topology.focusNf || "PCF";
  const nodes = topology.nodes.map((n) => ({
    id: n.id,
    label: n.label,
    color:
      n.id === focusNf
        ? { background: "#f0c040", border: "#c9a020" }
        : { background: "#4a9eff", border: "#2a7fd4" },
    font: { color: n.id === focusNf ? "#1a1d23" : "#fff", size: 14 },
    shape: "box",
  }));
  const edges = (topology.edges || []).map((e) => ({
    from: e.from,
    to: e.to,
    arrows: "to",
    title: (e.examples || []).join(", "),
  }));

  const network = new vis.Network(el, { nodes: new vis.DataSet(nodes), edges: new vis.DataSet(edges) }, {
    physics: { enabled: true, stabilization: { iterations: 120 } },
    interaction: { hover: true },
  });
  network.once("stabilizationIterationsDone", () => network.fit({ animation: true }));
}

function renderStory(data) {
  const spec = data.spec || {};
  const narrative = data.narrative || {};
  const ts = spec.tsNumber || tsFromPath();

  updatePlatformLinks(ts);

  document.title = `TS ${ts} — Spec Story`;
  document.getElementById("bc-ts").textContent = `TS ${ts}`;
  const displayTitle = narrative.title || spec.title || "Spec Story";
  const cleanTitle = displayTitle.startsWith("(referenced") ? narrative.tagline || "Spec Story" : displayTitle;
  document.getElementById("story-title").textContent = `TS ${ts} — ${cleanTitle}`;
  document.getElementById("story-tagline").textContent = narrative.tagline || narrative.summary || "";

  const banner = document.getElementById("version-banner");
  if (data.versionBanner) {
    banner.textContent = data.versionBanner;
    banner.classList.remove("hidden");
  } else {
    banner.classList.add("hidden");
  }

  const isStitchPage = Boolean(narrative.stitchAnchor);
  const isAnchor = ts === "29.512";
  const e2eFlow = data.enrichedE2eFlow || narrative.e2eFlow;
  const relatedLinks = narrative.relatedStory?.length
    ? narrative.relatedStory
    : (narrative.relatedSpecs || []).map((r) => ({ ts: r.ts, role: r.role }));
  const main = document.getElementById("story-main");

  const section = (id, title, desc, html, show = true) =>
    show && html
      ? `<section class="story-section" id="${id}"><h2>${title}</h2>${desc ? `<p>${desc}</p>` : ""}${html}</section>`
      : "";

  main.innerHTML = `
    <div class="meta-row">
      <span class="meta-chip">YAML <strong>${escapeHtml(data.versions?.yaml || "—")}</strong></span>
      <span class="meta-chip">PDF <strong>${escapeHtml(data.versions?.pdf || "—")}</strong></span>
      ${isAnchor ? `<span class="meta-chip">Stitch <strong>${data.storyGaps?.meta?.ready || 0}/${data.storyGaps?.meta?.total || 0}</strong> specs</span>` : ""}
      ${!isStitchPage && !isAnchor ? `<span class="meta-chip">Referenced by <strong>${data.referencedBy?.length || 0}</strong> YAMLs</span>` : ""}
      ${isAnchor || isStitchPage ? `<a class="master-journey-link" href="/stories">Master journey →</a>` : ""}
    </div>

    ${data.specReadiness && !isAnchor && isDebugMode() ? renderSpecReadiness(data.specReadiness) : ""}

    ${section(
      "sec-overview",
      isStitchPage || isAnchor ? (isAnchor ? "Overview" : "Supporting spec") : "Overview",
      "",
      `<p>${escapeHtml(narrative.summary || narrative.tagline || "")}</p>${narrative.whyItMatters ? `<p>${escapeHtml(narrative.whyItMatters)}</p>` : ""}${narrative.defines?.length ? `<ul>${narrative.defines.map((d) => `<li>${escapeHtml(d)}</li>`).join("")}</ul>` : ""}${narrative.primaryNf ? `<p class="meta-chip">NF: <strong>${escapeHtml(narrative.primaryNf)}</strong></p>` : ""}${renderRelatedStory(relatedLinks)}${isStitchPage ? `<p><a href="/stories">← Master story</a> · <a href="/specs?ts=29.512&view=story">TS 29.512 in platform</a></p>` : ""}${isAnchor ? `<p><a class="master-journey-link prominent" href="/stories">Open master journey</a></p>` : ""}`
    )}

    ${section(
      "sec-entry",
      "1. Entry points",
      isAnchor
        ? "Where session management policy control starts — SBI APIs and procedure triggers."
        : "How this spec enters the 512 story — procedures and SBI entry points.",
      renderEntryPoints(narrative.entryPoints),
      narrative.entryPoints?.length
    )}

    ${section(
      "sec-elements",
      "2. Major elements",
      isAnchor ? "Network functions involved in the 29.512 story." : "Network functions in this spec's story.",
      renderElements(narrative.elements, data.elementReadiness, isAnchor ? data.nfGaps : null),
      narrative.elements?.length
    )}

    ${section(
      "sec-interactions",
      "3. How elements interact",
      "Service relationships and operations relevant to this story.",
      `${renderInteractions(narrative.interactions, isAnchor ? data.operations : [])}<div id="interaction-graph"></div>`,
      narrative.interactions?.length || (isAnchor && data.operations?.length)
    )}

    ${section(
      "sec-gaps",
      "Story stitch — no more gaps",
      "Supporting specs that connect 512 to the live network. PDF and YAML sources stitched into the call flow below.",
      renderStoryGaps(data.storyGaps),
      isAnchor && data.storyGaps?.items?.length
    )}

    ${section(
      "sec-flow",
      "4. End-to-end call flow",
      isAnchor
        ? "PDU session establishment — green marker = steps defined in TS 29.512."
        : "Call flow for this spec within the 512 story — green marker = steps owned by this spec.",
      window.FlowRenderer.renderFlow(e2eFlow, ts, { nfMap: nfMapFromReadiness(data.elementReadiness) }),
      e2eFlow?.steps?.length
    )}

    ${section(
      "sec-operations",
      "API operations (YAML corpus)",
      "",
      renderOperations(data.operations),
      data.operations?.length
    )}

    ${section(
      "sec-refs",
      "Who references 29.512",
      "",
      renderReferencedBy(data.referencedBy),
      isAnchor && data.referencedBy?.length
    )}

    ${section(
      "sec-clauses",
      "PDF clause index",
      "",
      renderClauses(data.pdfClauses, data.highlightClauses || narrative.highlightClauses, ts),
      data.pdfClauses?.clauses?.length
    )}

    <div class="story-links">
      ${spec.url ? `<a href="${escapeHtml(spec.url)}" target="_blank" rel="noopener">3GPP archive</a>` : ""}
      <a href="/specs?ts=${encodeURIComponent(ts)}&view=map">Specs platform (map)</a>
      <a href="/specs?ts=${encodeURIComponent(ts)}&view=story">Specs platform (story)</a>
      ${!isAnchor ? `<a href="/specs?ts=29.512&view=story">TS 29.512 anchor</a>` : ""}
      <a href="/topology">NF Topology</a>
      ${data.yamls?.[0] ? `<a href="/?yaml=${encodeURIComponent(data.yamls[0].sourceFile)}">Open in Explorer</a>` : ""}
    </div>
  `;

  document.querySelectorAll(".toc-link").forEach((link) => {
    const id = link.getAttribute("href")?.slice(1);
    link.style.display = id && document.getElementById(id) ? "" : "none";
  });

  if (data.topology?.nodes?.length && (isAnchor || narrative.interactions?.length)) {
    initTopologyGraph({ ...data.topology, focusNf: narrative.primaryNf || "PCF" });
  }
  initTocSpy();
}

async function elementReadinessFromPlatform(elements) {
  const res = await fetch("/api/specs/platform");
  if (!res.ok) return null;
  const platform = await res.json();
  const byTs = Object.fromEntries((platform.specs || []).map((s) => [s.tsNumber, s]));
  const items = elements.map((el) => {
    const primarySpec = el.primarySpec || "29.512";
    const ps = byTs[primarySpec] || {};
    const nfListed = (ps.nfServices || []).includes(el.nf);
    const hasSpec = Boolean(ps.hasPdf);
    const hasYaml = nfListed ? Boolean(ps.hasYaml) : Boolean(ps.hasYaml);
    const hasStory = Boolean(ps.hasNarrative);
    const ready = hasSpec && hasYaml && hasStory;
    return {
      nf: el.nf,
      primarySpec,
      hasSpec,
      hasYaml,
      hasStory,
      ready,
      state: ready ? "ready" : hasSpec || hasYaml || hasStory ? "partial" : "gap",
      platformUrl: `/specs?ts=${encodeURIComponent(primarySpec)}`,
    };
  });
  const ready = items.filter((i) => i.ready).length;
  return { items, meta: { total: items.length, ready, complete: ready === items.length } };
}

async function ensureElementReadiness(data, ts) {
  if (data.elementReadiness?.items?.length) return data.elementReadiness;
  const elements = data.narrative?.elements;
  if (!elements?.length) return null;

  try {
    const res = await fetch(`/api/specs/${encodeURIComponent(ts)}/element-readiness`);
    if (res.ok) return res.json();
  } catch {
    /* try platform fallback */
  }

  return elementReadinessFromPlatform(elements);
}

async function loadStory() {
  if (isEmbedMode()) {
    document.body.classList.add("embed-mode");
  }

  const ts = tsFromPath();
  if (!ts) {
    document.getElementById("story-main").innerHTML = '<p class="error">No spec number in URL.</p>';
    return;
  }
  try {
    const res = await fetch(`/api/specs/${encodeURIComponent(ts)}/detail`);
    if (!res.ok) throw new Error(res.status === 404 ? "Spec not found" : res.statusText);
    const data = await res.json();
    data.elementReadiness = await ensureElementReadiness(data, ts);
    renderStory(data);
  } catch (err) {
    document.getElementById("story-main").innerHTML = `<p class="error">${escapeHtml(err.message)}</p>`;
  }
}

window.addEventListener("load", loadStory);

File: backend/services/spec_story.py
import json
import re

from backend.config import BANK_DIR, YAML_DIR
from backend.services.nf_topology import get_topology
from backend.services.pdf_clauses import load_pdf_clauses, pdf_metadata
from backend.services.spec_narrative import load_narrative, normalize_ts_number
from backend.services.story_stitch import (
    _STITCH_TITLES,
    build_element_readiness,
    build_nf_gaps,
    build_spec_readiness,
    build_story_gaps,
    enrich_e2e_flow,
    minimal_narrative_for_stitch,
)
from backend.services.ts_catalog import get_ts_catalog
from backend.services.nf_map import build_spec_map

TS_VERSION_RE = re.compile(r"TS\s+(\d+\.\d+)\s+V([\d.]+)", re.I)


def _find_spec_entry(ts_number: str) -> dict | None:
    for spec in build_spec_map().get("specs", []):
        if spec.get("tsNumber") == ts_number:
            return spec
    return None


def _catalog_entry(ts_number: str) -> dict | None:
    for spec in get_ts_catalog().get("specs", []):
        if spec.get("tsNumber") == ts_number:
            return spec
    return None


def _yaml_version(ts_number: str, yamls: list[dict]) -> str:
    for entry in yamls:
        path = YAML_DIR / entry.get("sourceFile", "")
        if not path.exists():
            continue
        text = path.read_text(encoding="utf-8", errors="ignore")[:4000]
        match = TS_VERSION_RE.search(text)
        if match and match.group(1) == ts_number:
            return f"V{match.group(2)}"
    return ""


def _load_operations(source_files: list[str]) -> list[dict]:
    operations: list[dict] = []
    for source in source_files:
        bank_name = source.replace(".yaml", "_Bank.json")
        bank_path = BANK_DIR / bank_name
        if not bank_path.exists():
            continue
        with open(bank_path, "r", encoding="utf-8") as f:
            bank = json.load(f)
        for msg in bank.get("messages", []):
            operations.append(
                {
                    "sourceFile": source,
                    "service": bank.get("service", ""),
                    "nf": bank.get("nf", ""),
                    "path": msg.get("path", ""),
                    "method": msg.get("method", ""),
                    "operationId": msg.get("operationId", ""),
                    "summary": msg.get("summary", ""),
                }
            )
    return operations


def _resolve_spec_title(ts_number: str, spec: dict | None, catalog: dict | None, narrative: dict | None) -> str:
    if narrative and narrative.get("title"):
        return narrative["title"]
    title = ""
    if spec:
        title = spec.get("title", "")
    elif catalog:
        title = catalog.get("title", "")
    if title.startswith("(referenced"):
        title = _STITCH_TITLES.get(ts_number, "")
    return title


def _topology_slice(center_nf: str) -> dict:
    topo = get_topology(min_weight=1, core_only=False)
    nodes = {n["id"] for n in topo.get("nodes", [])}
    if center_nf not in nodes:
        return {"nodes": [], "edges": []}

    neighbor_ids = {center_nf}
    edges: list[dict] = []
    for edge in topo.get("edges", []):
        if edge["from"] == center_nf or edge["to"] == center_nf:
            neighbor_ids.add(edge["from"])
            neighbor_ids.add(edge["to"])
            edges.append(edge)

    nodes_out = [n for n in topo.get("nodes", []) if n["id"] in neighbor_ids]
    return {"nodes": nodes_out, "edges": edges}


def _layer_hints(ts_number: str, narrative: dict) -> dict:
    architecture_ts = ""
    for rel in narrative.get("relatedSpecs", []):
        ts = rel.get("ts", rel) if isinstance(rel, dict) else rel
        if str(ts).startswith("23."):
            architecture_ts = normalize_ts_number(str(ts))
            break
    if not architecture_ts:
        for ep in narrative.get("entryPoints", []):
            ref = ep.get("refSpec", "")
            if ref.startswith("23."):
                architecture_ts = normalize_ts_number(ref)
                break

    operations: list[dict] = []
    for ep in narrative.get("entryPoints", []):
        op_id = ep.get("operationId") or ep.get("name", "")
        if not op_id:
            continue
        operations.append(
            {
                "operationId": op_id,
                "explorerUrl": f"/?q={op_id}",
                "description": ep.get("description", ""),
            }
        )
        if len(operations) >= 3:
            break

    return {
        "architectureTs": architecture_ts,
        "serviceTs": ts_number,
        "operations": operations,
    }


def _preview_steps(narrative: dict, primary_nf: str) -> list[dict]:
    steps = narrative.get("e2eFlow", {}).get("steps", [])
    primary_nf = primary_nf or ""
    filtered = [s for s in steps if primary_nf and primary_nf in (s.get("actors") or [])]
    if not filtered:
        filtered = steps[:4]
    else:
        filtered = filtered[:4]
    return filtered


def build_spec_preview(ts_number: str) -> dict | None:
    ts_number = normalize_ts_number(ts_number)
    narrative = load_narrative(ts_number)
    if not narrative:
        narrative = minimal_narrative_for_stitch(ts_number, "29.512")
    if not narrative:
        return None

    primary_nf = narrative.get("primaryNf", "")
    entry_points = (narrative.get("entryPoints") or [])[:3]
    defines = (narrative.get("defines") or [])[:2]

    return {
        "tsNumber": ts_number,
        "title": narrative.get("title", f"TS {ts_number}"),
        "tagline": narrative.get("tagline", ""),
        "summary": narrative.get("summary", narrative.get("tagline", "")),
        "whyItMatters": narrative.get("whyItMatters", ""),
        "defines": defines,
        "primaryNf": primary_nf,
        "entryPoints": entry_points,
        "e2eFlowSteps": _preview_steps(narrative, primary_nf),
        "layerHints": _layer_hints(ts_number, narrative),
        "storyUrl": f"/specs/{ts_number}",
        "platformUrl": f"/specs?ts={ts_number}",
        "stitchAnchor": narrative.get("stitchAnchor"),
    }


def build_spec_detail(ts_number: str, refresh_pdf: bool = False) -> dict | None:
    ts_number = normalize_ts_number(ts_number)
    spec = _find_spec_entry(ts_number)
    catalog = _catalog_entry(ts_number)
    pdf_info = pdf_metadata(ts_number)
    if not spec and not catalog and not pdf_info:
        return None

    narrative = load_narrative(ts_number)
    stitch_anchor = "29.512"
    if not narrative:
        narrative = minimal_narrative_for_stitch(ts_number, stitch_anchor)
    yamls: list[dict] = []
    if spec:
        for nf_block in spec.get("nfs", []):
            yamls.extend(nf_block.get("yamls", []))

    source_files = sorted({y.get("sourceFile", "") for y in yamls if y.get("sourceFile")})
    operations = _load_operations(source_files)

    referenced_by = []
    if catalog:
        referenced_by = catalog.get("referenceOnlyIn", [])

    resolved_title = _resolve_spec_title(ts_number, spec, catalog, narrative)
    primary_nf = (narrative or {}).get("primaryNf") or (
        spec["nfs"][0]["nf"] if spec and spec.get("nfs") else "PCF"
    )
    topology = _topology_slice(primary_nf)

    pdf_clauses = load_pdf_clauses(ts_number, refresh=refresh_pdf)
    yaml_version = _yaml_version(ts_number, yamls)

    version_banner = None
    if pdf_info and yaml_version and pdf_info.get("version") and pdf_info["version"] != yaml_version:
        version_banner = (
            f"PDF is {pdf_info['version']}; YAML corpus is {yaml_version}. "
            "API list reflects YAML; narrative and clauses reflect PDF."
        )

    spec_out = dict(spec) if spec else {
        "tsNumber": ts_number,
        "title": "",
        "url": (catalog.get("urls") or [""])[0] if catalog else "",
        "versions": catalog.get("versions", []) if catalog else [],
        "nfs": [],
        "yamlCount": 0,
    }
    if resolved_title:
        spec_out["title"] = resolved_title

    return {
        "spec": spec_out,
        "catalog": {
            "referenceCount": catalog.get("referenceCount", 0) if catalog else 0,
            "referenceOnly": catalog.get("referenceOnly", False) if catalog else False,
        },
        "narrative": narrative,
        "pdf": pdf_info,
        "pdfClauses": pdf_clauses,
        "yamls": yamls,
        "operations": operations,
        "referencedBy": referenced_by,
        "topology": topology,
        "versions": {
            "yaml": yaml_version,
            "pdf": pdf_info.get("version", "") if pdf_info else "",
        },
        "versionBanner": version_banner,
        "relatedSpecs": narrative.get("relatedSpecs", []) if narrative else [],
        "storyGaps": build_story_gaps(stitch_anchor) if ts_number == stitch_anchor else None,
        "elementReadiness": (
            build_element_readiness(narrative.get("elements", []), stitch_anchor)
            if ts_number == stitch_anchor and narrative
            else None
        ),
        "nfGaps": (
            build_nf_gaps(narrative.get("elements", []), stitch_anchor)
            if ts_number == stitch_anchor and narrative
            else None
        ),
        "enrichedE2eFlow": (
            enrich_e2e_flow(ts_number, stitch_anchor)
            if narrative and narrative.get("e2eFlow")
            else None
        ),
        "specReadiness": build_spec_readiness(ts_number, narrative) if narrative else None,
        "stitchAnchor": narrative.get("stitchAnchor") if narrative else None,
        "highlightClauses": narrative.get("highlightClauses", []) if narrative else [],
    }
